src/cpu/testers/traffic_gen/gups_gen.cc

```diff
@@ -0,0 +1,444 @@
+        stats.totalUpdates++;
+        DPRINTF(GUPSGen, "%s: onTheFlyRequests: %d.\n",
+                __func__, onTheFlyRequests);
+
+        stats.totalWrites++;
+        stats.totalBytesWritten += elementSize;
+        stats.totalWriteLat += curTick() - exitTimes[pkt->req];
+
+        exitTimes.erase(pkt->req);
+        delete pkt;
+
+        if ((stats.totalUpdates.value() >= updateLimit) && updateLimit != 0) {
+            exitSimLoop(name() + " has encountered the exit state and will "
+                "terminate the simulation.\n");
+            return;
+        }
+    }
+    else {
+        DPRINTF(GUPSGen, "handleResponse: received a read resp.\n");
+
+        stats.totalReads++;
+        stats.totalBytesRead += elementSize;
+        stats.totalReadLat += curTick() - exitTimes[pkt->req];
+
+        exitTimes.erase(pkt->req);
+
+        responsePool.push(pkt);
+    }
+    if (doneReading && requestPool.empty() && onTheFlyRequests == 0){
+        exitSimLoop(name() + " has encountered the exit state and will "
+                "terminate the simulation.\n");
+        return;
+    }
+    if ((requestPool.size() < reqQueueSize) &&
+        (!doneReading || !responsePool.empty())) {
+        if (!nextCreateEvent.scheduled()) {
+            schedule(nextCreateEvent, nextCycle());
+        }
+    }
+}
+
+void GUPSGen::wakeUp()
+{
+    if (!nextSendEvent.scheduled() && !requestPool.empty())
+    {
+        schedule(nextSendEvent, nextCycle());
+    }
+}
+
+void
+GUPSGen::createNextReq() {
+
+    if (!responsePool.empty()) {
+        PacketPtr pkt = responsePool.front();
+        responsePool.pop();
+
+        uint64_t *updated_value = pkt->getPtr<uint64_t>();
+        *updated_value ^= updateTable[pkt->req];
+        updateTable.erase(pkt->req);
+
+        uint8_t write_data[elementSize];
+        std::memcpy(write_data, updated_value, elementSize);
+        Addr addr = pkt->getAddr();
+        delete pkt;
+
+        PacketPtr new_pkt = getWritePacket(addr, elementSize, write_data);
+        requestPool.push(new_pkt);
+    }
+    else if (!doneReading) {
+        assert (rIndex < 128);
+
+        randomSeeds[rIndex] = (randomSeeds[rIndex] << 1) ^
+                        ((int64_t) randomSeeds[rIndex] < 0 ? POLY : 0);
+        uint64_t value = randomSeeds[rIndex];
+        uint64_t index = randomSeeds[rIndex] & (tableSize - 1);
+        Addr addr = indexToAddr(index);
+        PacketPtr pkt = getReadPacket(addr, sizeof(uint64_t));
+        updateTable[pkt->req] = value;
+        requestPool.push(pkt);
+
+        rIndex++;
+        if (rIndex >= 128) {
+            rIndex = 0;
+            iterations++;
+            if (iterations >= numUpdates / 128) {
+                DPRINTF(GUPSGen, "sendNextReq: Done creating reads.\n");
+                doneReading = true;
+            }
+        }
+    }
+
+    if (!nextCreateEvent.scheduled()) {
+        if ((requestPool.size() < reqQueueSize) &&
+            (!doneReading || !responsePool.empty())) {
```

Comment:
You create another if the not done reading *or* the response pool is empty?

I see why you're doing this now... This confusing condition is evidence this should be two functions. You should have one function for generating reads, and when you get a response you should immediately create a write (i.e., in the same function call chain).

Reply:
The reason I decided to use one function to create both reads and writes was to have control over when a write is issued (to prevent cases that a read response is received and the requestPool is full), I could separate the functions for creating writes and reads. Or I can create as many writes as I can (empty out the responsePool) in the createNextReq function.